internal/host/hostcommands/instruction_manager.go

```diff
@@ -178,10 +190,69 @@ func (i *InstructionManager) GetNextSteps(ctx context.Context, host *models.Host
 	} else {
 		returnSteps.NextInstructionSeconds = defaultNextInstructionInSec
 	}
+
+	// If the agent isn't compatible with the service and the host is in a state that allows to
+	// upgrade the agent, then replace all the calculated steps with a single step to upgrade
+	// the agent:
+	if i.isAgentUpgradeAllowed(host) &&
+		!common.IsAgentCompatible(i.config.AgentImage, host.DiscoveryAgentVersion) {
+		log.WithFields(logrus.Fields{
+			"expected_image": i.config.AgentImage,
+			"actual_image":   host.DiscoveryAgentVersion,
+		}).Info(
+			"Agent image ins't compatible with the service, and it can be upgraded, " +
+				"will replace all the calculated steps with a single step to " +
+				"upgrade the image",
+		)
+		var err error
+		returnSteps.Instructions, err = i.upgradeAgentCmd.GetSteps(ctx, host)
+		if err != nil {
+			return models.Steps{}, err
+		}
+		returnSteps.PostStepAction = swag.String(models.StepsPostStepActionContinue)
+		returnSteps.NextInstructionSeconds = defaultNextInstructionInSec
+		events.SendUpgradeAgentStartedEvent(
+			ctx,
+			i.eventsHandler,
+			*host.ID,
+			hostutil.GetHostnameForMsg(host),
+			host.InfraEnvID,
+			host.ClusterID,
+			i.config.AgentImage,
+		)
+	}
+
 	logSteps(returnSteps, InfraEnvID, hostID, log)
 	return returnSteps, nil
 }
 
+// isAgentUpgradeAllowed checks if the current state of the host allows the agent to be upgraded.
+// For example, it is not allowed to upgrade the agent when the installation of the cluster is in
+// progress.
+func (i *InstructionManager) isAgentUpgradeAllowed(host *models.Host) bool {
+	if !i.config.EnableUpgradeAgent {
+		return false
+	}
+	if i.isStepDisabled(models.StepTypeUpgradeAgent) {
+		return false
+	}
+	switch *host.Status {
+	case models.HostStatusBinding,
+		models.HostStatusDisabled,
+		models.HostStatusDisabledUnbound,
+		models.HostStatusDiscovering,
+		models.HostStatusDiscoveringUnbound,
+		models.HostStatusInsufficient,
+		models.HostStatusInsufficientUnbound,
+		models.HostStatusKnown,
+		models.HostStatusKnownUnbound,
+		models.HostStatusPendingForInput:
```

Comment:
The thing I find strange in some of these is that somesutatuses are not possible for agent upgrade..
If the agent requires an upgrade the host status can't be known(+Unbound) / discovering(+Unbound) because if we are upgrading the agent it means it has a failing validation.
So I'd expect the host status to be insufficient/insufficientUnbound.
I can understand disabled* though I'm not sure if we should upgrade the agent on a disabled host.
Unsure what state HostStatusPendingForInput represents

Reply:
The decision to send the upgrade-agent command is not a consequence of the failure of the validation. Both use the same condition, but are calculated independently and at different times. It can happen that the validation has not yet been evaluated, but still the decision will be made to send the upgrade-agent command. So the host can be in the "discovering" state (for example, same for the other states) and the service will make the decision to send the "upgrade-agent" command.

I will remove the disabled and pending for input states.